zebra-state/src/service.rs

```diff
@@ -270,6 +270,105 @@ impl StateService {
             state: IterState::NonFinalized(hash),
         }
     }
+
+    /// Return a list of block hashes in the best chain, following the first matching hash in `known_blocks`.
+    ///
+    /// Starts from the first matching hash in the best chain, ignoring all other hashes in `known_blocks`.
+    /// Only matches and returns hashes from the best chain, including non-finalized blocks.
+    /// Stops the list of hashes after:
+    ///   * adding the non-finalized best tip,
+    ///   * adding the stop hash to the list, if it is in the best chain, or
+    ///   * adding 500 hashes to the list.
+    pub fn find_chain_hashes(
+        &self,
+        known_blocks: Vec<block::Hash>,
+        stop: Option<block::Hash>,
+    ) -> Vec<block::Hash> {
+        let mut res: Vec<block::Hash> = Vec::new();
+
+        // Known blocks can't be empty
+        if known_blocks.is_empty() {
+            return res;
+        }
+
+        // Get chain tip
+        let (chain_tip_height, ..) = self.tip().expect("tip must always be available");
+
+        // Get requested tip
+        let locator_tip_hash = known_blocks.first().expect("we must have a tip hash");
+        let locator_tip_height = self
+            .height_by_hash(*locator_tip_hash)
+            .expect("tip hash must have a height");
+        let chain = self.chain(*locator_tip_hash);
+
+        // Requested tip must be below our chain tip
+        if locator_tip_height >= chain_tip_height {
+            return res;
+        }
```

Comment:
This case should be impossible, because we should look up the locator tip in the best chain, and by definition every block in the best chain is at or below the best chain tip.